server/src/main/java/org/zanata/magpie/service/PersistentTranslationService.java

```diff
@@ -182,15 +202,34 @@ public PersistentTranslationService(TextFlowDAO textFlowDAO,
             // see if we already have a matched text flow
             // (either in the same document or copied from other document)
             TextFlow tf = indexTextFlowMap.get(indexes.iterator().next());
-            if (tf == null) {
-                tf = createTextFlow(document, source, fromLocale);
+
+            try {
+                if (tf == null) {
+                    tf = createTextFlow(document, source, fromLocale);
+                }
+                requestedTextFlows.add(tf.getContentHash());
+                TextFlowTarget target =
+                        new TextFlowTarget(translation.getPlainTranslation(),
+                                translation.getRawTranslation(), tf,
+                                toLocale, backendID);
+                createOrUpdateTextFlowTarget(target);
+            } catch (Exception e) {
+                List<Throwable> causalChain = Throwables.getCausalChain(e);
+                Optional<Throwable> duplicateKeyEx = causalChain.stream()
+                        .filter(t -> t instanceof PSQLException &&
+                                t.getMessage().contains(
+                                        "ERROR: duplicate key value violates unique constraint"))
+                        .findAny();
+                if (duplicateKeyEx.isPresent()) {
+                    LOG.warn("concurrent requests for document {}", document.getUrl());
+                    // we ignore the failed update
+                }
```

Comment:
So we are not handling the concurrent requests? just return error?

Reply:
we don't use lock anymore so this will happen. But we will record the MT triggering.

Reply:
Record it via the log message you mean? But the idea is to record all our translation requests in the database (even the duplicate ones), so that they can be reconciled with the bill from the translation service.

Reply:
Record it in database. There is a new table for it.